ingestor/service.go

```diff
@@ -381,17 +395,29 @@ func (s *Service) HandleTransfer(w http.ResponseWriter, r *http.Request) {
 	w.WriteHeader(http.StatusAccepted)
 }
 
-func (s *Service) UploadSegments() error {
+func (s *Service) Shutdown(ctx context.Context) error {
+	if err := s.metrics.Close(); err != nil {
+		return err
+	}
+
+	if err := s.UploadSegments(ctx); err != nil {
+		return fmt.Errorf("Failed to upload segments: %s", err.Error())
+	}
+
+	return nil
+}
+
+func (s *Service) UploadSegments(ctx context.Context) error {
 	if err := s.batcher.BatchSegments(); err != nil {
 		return err
 	}
 	logger.Infof("Waiting for upload queue to drain, %d batches remaining", len(s.uploader.UploadQueue()))
 	logger.Infof("Waiting for transfer queue to drain, %d batches remaining", len(s.replicator.TransferQueue()))
 
+	timeoutCtx, cancel := context.WithTimeout(ctx, shutdownTimeout)
```

Comment:
Not quite what I was getting at.  I was suggesting to move this part out of `UploadSegments` to let the caller define when to timeout by passing a  context.  So move L417-L418 to the call site.

Reply:
Gotcha. updated to pass the timeout context from the shutdown runner.